include/deal.II/dofs/dof_vector.h

```diff
@@ -0,0 +1,243 @@
+   * this object and released when the object itself expires.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints);
+
+  /**
+   * Construct an object with a DoFHandler but without constraints and
+   * borrow a vector from GrowingVectorMemory for its data. This
+   * vector is then owned by this object and released when the object
+   * itself expires.
+   *
+   * @note This constructor uses the ConstraintMatrix
+   * #no_constraints. The created object thus can output vectors in
+   * distributed format, but writing into vectors will most likely
+   * fail without using a constraint matrix from somewhere else.
+   */
+  DoFVector (const DH &dh);
+
+  /**
+   * Construct an object with a DoFHandler and borrow a vector from
+   * GrowingVectorMemory for its data. This vector is not owned by the
+   * object and read-access only.
+   *
+   * @note An object created this way is effectively constant. In
+   * order to avoid obscure access errors to the data vector, it
+   * should also be declared `const`.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints, const VECTOR &v);
+
+  /**
+   * Construct an object with a DoFHandler but without constraints and
+   * borrow a vector from GrowingVectorMemory for its data. This
+   * vector is not owned by the object and read-access only.
+   *
+   * @note An object created this way is effectively constant. In
+   * order to avoid access errors to the data vector, it should also
+   * be declared `const`. Also, vectors are expected to be in
+   * distributed format, since the information on constraints is
+   * missing.
+   */
+  DoFVector (const DH &dh, const VECTOR &v);
+
+  /**
+   * Destructor, deleting own data.
+   */
+  ~DoFVector ();
+
+  /**
+   * Modify the vector dimensions to match the current state of the
+   * dof handler.
+   */
+  void sync ();
+
+  ///@}
+
+  /// @name Access to members
+  ///@{
+  /**
+   * Access to the DoFHandler
+   */
+  const DH &dof_handler () const;
+
+  /**
+   * Read-only access to the data vector
+   */
+  const VECTOR &data () const;
+
+  /**
+   * Read-write access to the data vector.
+   *
+   * Read-write access is only possible, if the vector is owned by the
+   * object.
+   */
+  VECTOR &data ();
+  ///@}
+
+private:
+  /**
+   * The pool for allocating the vector.
+   */
+  GrowingVectorMemory<VECTOR> mem;
+
+  /**
+   * Empty constrainty matrix for constructors without one.
+   */
+  ConstraintMatrix no_constraints;
+
+  /**
+   * Pointer to the dof handler.
+   */
+  SmartPointer<const DH, DoFVector<DH, VECTOR> > dh;
+
+  /**
+   * Pointer to the dof handler.
+   */
+  SmartPointer<const ConstraintMatrix, DoFVector<DH, VECTOR> > constraints;
+
+  /**
+   * The actual data vector, if owned by this object.
+   */
+  VECTOR *my_data;
+
+  /**
+   * The data vector if referencing a vector not owned by this object.
+   */
+  SmartPointer<const VECTOR, DoFVector<DH, VECTOR> > other_data;
```

Comment:
Am I correct that either `my_data` or `other_data` are always NULL? If so, maybe add a comment.